client/src/util/is-recursively-primitive.js

```diff
@@ -0,0 +1,45 @@
+import isPlainObject from 'is-plain-object'
+
+// Whether an object is all primitives all the way down. We consider
+// functions non-primitives, lump Dates and ArrayBuffers into
+// primitives, and objects and arrays that contain only recursively
+// primitive items to be primitives.
+export function isRecursivelyPrimitive(value) {
+  if (value === null) {
+    return true
+  }
+  if (value === undefined) {
+    return false
+  }
+  if (typeof value === 'function') {
+    return false
+  }
+  if ([ 'boolean', 'number', 'string' ].indexOf(typeof value) !== -1) {
+    return true
+  }
+  if (value instanceof Date || value instanceof ArrayBuffer) {
+    return true
+  }
+  const isPOJO = isPlainObject(value)
+  const isArray = Array.isArray(value)
+  if (!isPOJO && !isArray && typeof value === 'object') {
```

Comment:
Yea only difference is we'd now return `ArrayTerm<PrimitiveTerm>` instead of `PrimitiveTerm`, which seems fine if not expected.